src/cudamatrix/cu-kernels.cu

```diff
@@ -966,6 +966,168 @@ static void _trace_mat_mat(const Real* A, const Real* B, MatrixDim dA,
 
 }
 
+#if __CUDA_ARCH__ < 600
+template <typename Type>
+inline __device__ void myAtomicAdd(Type *address, Type val) {
+  atomicAdd(address, val);
+}
+// Ref:
+// http://on-demand.gputechconf.com/gtc/2013/presentations/S3101-Atomic-Memory-Operations.pdf
+template <>
+inline __device__ void myAtomicAdd(double *address, double val) {
+  unsigned long long int *address_as_ull = (unsigned long long int *)address;
+  unsigned long long int old = *address_as_ull, assumed;
+  do {
+    assumed = old;
+    old = atomicCAS(address_as_ull, assumed,
+                    __double_as_longlong(val + __longlong_as_double(assumed)));
+  } while (assumed != old);
+}
+#else
+#define myAtomicAdd(a, b) atomicAdd(a, b)
+#endif  // __CUDA_ARCH__
+
+// Kernel to compute tr(AB^T) and store result in GPU memory.
+// Each block computes a partial sum and adds to result atomicaly.
+// Order of computation is not guaranteed, but this doesn't affect
+// accuracy significantly.
+template <typename Real, int THREADS_X, int THREADS_Y, int unroll_count>
+__global__ void _trace_mat_mat_trans_atomic(Real * result,
+                                            const Real * A, const Real * B,
+                                            MatrixDim dA, int B_stride) {
+  // This kernel assumes result is already set to zero
+  Real thread_data = static_cast<Real>(0);
+  int colStart = blockIdx.x * blockDim.x + threadIdx.x;
+
+  // Thread reduction while reading from global memory
+  // in a grid strided loop
+  if (colStart < dA.cols) {
+    int rowStart = blockIdx.y * blockDim.y + threadIdx.y;
+    int stride = blockDim.y * gridDim.y;
+
+    int unroll_stride = unroll_count * stride;
+    // Closest multiple of unroll_stride less than dA.rows
+    int Nmod = (dA.rows / unroll_stride) * unroll_stride;
+    int j = rowStart;
+    for (; j < Nmod; j += unroll_stride) {
+      Real A_vals[unroll_count];
+      Real B_vals[unroll_count];
+
+      #pragma unroll
+      for (int u = 0; u < unroll_count; ++u) {
+        int idx_A = colStart + (j + u*stride) * dA.stride;
+        int idx_B = colStart + (j + u*stride) * B_stride;
+        A_vals[u] = A[idx_A];
+        B_vals[u] = B[idx_B];
+      }
+
+      #pragma unroll
+      for (int u = 0; u < unroll_count; ++u) {
+        thread_data += A_vals[u] * B_vals[u];
```

Comment:
I'm curious why you made this a separate loop, why is this faster than just doing, in the loop, thread_data += A[idx_A] + B[idx_B] ?

Reply:
Assuming no compiler optimizations, having only one loop with `thread_data += A[idx_A] * B[idx_B]` would mean that you would have sequential dependencies between the load and the compute instructions. Having them in two separate loops means you have multiple independent loads into separate registers followed by multiple independent compute ops that only depend on the corresponding load op thus allowing for better instruction level parallelism. Of course, compilers are much smarter and can usually do this automatically especially with `#pragma unroll`, but I typically like to structure the code so it helps the compiler and also makes it easier to understand the performance reasoning for unrolling this loop.